File: capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int D) {
        int tot=0,minn=0;
        for(int i=0;i<weights.size();i++)
            tot+=weights[i], minn=max(minn,weights[i]);
        int ans=tot;
        while(tot>=minn){
            int mid=(tot+minn)/2;
            //cout<<minn<<" "<<tot<<" "<<mid<<endl;
            int d=1, sum=0;
            for(int i=0;i<weights.size();i++){
                sum+=weights[i];
                if(sum>mid){
                    sum-=weights[i];
                    sum=weights[i];
                    d++;
                }
                else if(sum==mid){
                    d++;
                    sum=0;
                }
            }
            if(sum>mid) d++;
            cout<<mid<<" "<<d<<endl;
            if(d>D){
                minn=mid+1;
            }
            else{
                ans=min(ans,mid);
                tot=mid-1;
            }
        }
        return ans;
    }
};

File: capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        
        int l=1,h=0;
        
        for(int i:weights){
            if(i>l)
                l=i;
            h+=i;
        }
        
        int res=h;
        
        while(h>=l){
            
            int m=l+(h-l)/2;
            
            int d=1, s=0;
            
            for(int i:weights){
                
                if(s+i>m){
                    d++;
                    s=i;
                }
                else{
                    s+=i;
                }
            }
            
            if(d>days){
                l=m+1;
            }
            else{
                h=m-1;
                res=m;
            }
        }
        
        return res;
    }
}